### PR TITLE
Add a few additional diags (divE etc.)

### PR DESCRIPTION
**Pull request by MaxThevenet**
This relatively small PR provides minor increments over previous one:
- Create sub-directories to re-organize a bit Diagnostics/
- Move `Average` and `Interpolate` to `Utils/`
- Allow for more plotting options (divE, etc.).

The new diagnostics can be used with
```
diagnostics.ndiags = 1
diagnostics.diags_names = d1

d1.diag_type = Full
d1.fields_to_plot = By Bz jx rho divE part_per_cell
d1.period = 4
# d1.file_prefix = diags/ehouais
```

### CODE REVIEW

**Comment by MaxThevenet** [comment]
This PR is ready for review. Unless I miss something, that would include all current relevant options for diagnostics. Once @EZoni and @RevathiJambunathan have implemented coarsening (and @ax3l or myself have implemented OpenPMD format for these new diags), I will make a final PR to
- Get rid of all old diagnostics
- Document the new ones on RTD
- update examples so that they use the new diags

**Comment by ax3l** [comment]
<3 subdirs!

File: Source/Diagnostics/ComputeDiagFunctors/DivBFunctor.H [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+class
+DivBFunctor : public ComputeDiagFunctor
+{
+public:
+    /** Constructor.
+     * \param[in] arr_mf_src source multifabs (3 elements for x y z).
+     * \param[in] lev level of multifab.
+     * \param[in] ncomp Number of component of mf_src to cell-center in dst multifab.
+     */
+    DivBFunctor(std::array<const amrex::MultiFab*, 3> arr_mf_src, int lev, int ncomp=1);
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/DivBFunctor.H]:
```suggestion
    DivBFunctor(const std::array<const amrex::MultiFab*, 3> arr_mf_src, const int lev, const int ncomp=1);
```

File: Source/Diagnostics/ComputeDiagFunctors/DivBFunctor.H [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    /** \brief Compute DivB directly into mf_dst.
+     *
+     * \param[out] mf_dst output MultiFab where the result is written
+     * \param[in] dcomp first component of mf_dst in which cell-centered
+     *            data is stored
+     */
+    virtual void operator()(amrex::MultiFab& mf_dst, int dcomp) const override final;
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/DivBFunctor.H]:
```suggestion
    virtual void operator()(amrex::MultiFab& mf_dst, const int dcomp) const override final;
```

File: Source/Diagnostics/ComputeDiagFunctors/DivBFunctor.cpp [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#include "DivBFunctor.H"
+#include "Utils/Average.H"
+
+using namespace amrex;
+
+DivBFunctor::DivBFunctor(std::array<const amrex::MultiFab*, 3> arr_mf_src, int lev, int ncomp)
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/DivBFunctor.cpp]:
```suggestion
DivBFunctor::DivBFunctor(const std::array<const amrex::MultiFab*, 3> arr_mf_src, const int lev, const int ncomp)
```

File: Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.H [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+class
+DivEFunctor : public ComputeDiagFunctor
+{
+public:
+    /** Constructor.
+     * \param[in] arr_mf_src source multifabs (3 elements for x y z).
+     * \param[in] lev level of multifab.
+     * \param[in] ncomp Number of component of mf_src to cell-center in dst multifab.
+     */
+    DivEFunctor(std::array<const amrex::MultiFab*, 3> arr_mf_src, int lev, int ncomp=1);
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.H]:
```suggestion
    DivEFunctor(const std::array<const amrex::MultiFab*, 3> arr_mf_src, const int lev, const int ncomp=1);
```

File: Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.H [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    /** \brief Compute DivE directly into mf_dst.
+     *
+     * \param[out] mf_dst output MultiFab where the result is written
+     * \param[in] dcomp first component of mf_dst in which cell-centered
+     *            data is stored
+     */
+    virtual void operator()(amrex::MultiFab& mf_dst, int dcomp) const override final;
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.H]:
```suggestion
    virtual void operator()(amrex::MultiFab& mf_dst, const int dcomp) const override final;
```

File: Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.cpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#include "DivEFunctor.H"
+#include "Utils/Average.H"
+
+using namespace amrex;
+
+DivEFunctor::DivEFunctor(std::array<const amrex::MultiFab*, 3> arr_mf_src, int lev, int ncomp)
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.cpp]:
```suggestion
DivEFunctor::DivEFunctor(const std::array<const amrex::MultiFab*, 3> arr_mf_src, const int lev, const int ncomp)
```

File: Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    : ComputeDiagFunctor(ncomp), m_arr_mf_src(arr_mf_src), m_lev(lev)
+{}
+
+void
+DivEFunctor::operator()(amrex::MultiFab& mf_dst, int dcomp) const
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.cpp]:
```suggestion
DivEFunctor::operator()(amrex::MultiFab& mf_dst, const int dcomp) const
```

File: Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.H (hunk 5 above)
Comment (ax3l) [reply Source/Diagnostics/ComputeDiagFunctors/DivEFunctor.H]: if you like you can also mark the whole class as `final`.

File: Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.H [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+PartPerCellFunctor : public ComputeDiagFunctor
+{
+public:
+    /** Constructor.
+     * \param[in] mf_src source multifab. Must be nullptr as not source MF is needed
+     *            to compute the number of particles per cell, banane.
+     * \param[in] lev level of multifab. Used for averaging in rz.
+     * \param[in] ncomp Number of component of mf_src to cell-center in dst multifab.
+     */
+    PartPerCellFunctor(amrex::MultiFab* mf_src, int lev, int ncomp=1);
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.H]:
```suggestion
    PartPerCellFunctor(const amrex::MultiFab* mf_src, const int lev, const int ncomp=1);
```

File: Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.H [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    /** \brief Compute the number of particles per cell directly into mf_dst.
+     *
+     * \param[out] mf_dst output MultiFab where the result is written
+     * \param[in] dcomp first component of mf_dst in which cell-centered
+     *            data is stored
+     */
+    virtual void operator()(amrex::MultiFab& mf_dst, int dcomp) const override final;
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.H]:
```suggestion
    virtual void operator()(amrex::MultiFab& mf_dst, const int dcomp) const override final;
```

File: Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.cpp [hunk 10]
```diff
@@ -0,0 +1,25 @@
+#include "PartPerCellFunctor.H"
+#include "WarpX.H"
+
+using namespace amrex;
+
+PartPerCellFunctor::PartPerCellFunctor(amrex::MultiFab* mf_src, int lev, int ncomp)
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.cpp]:
```suggestion
PartPerCellFunctor::PartPerCellFunctor(const amrex::MultiFab* mf_src, const int lev, const int ncomp)
```

File: Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.cpp [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    : ComputeDiagFunctor(ncomp), m_lev(lev)
+{
+    // mf_src will not be used, let's make sure it's null.
+    AMREX_ALWAYS_ASSERT(mf_src == nullptr);
+    // Write only in one output component.
+    AMREX_ALWAYS_ASSERT(ncomp == 1);
+}
+
+void
+PartPerCellFunctor::operator()(amrex::MultiFab& mf_dst, int dcomp) const
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerCellFunctor.cpp]:
```suggestion
PartPerCellFunctor::operator()(amrex::MultiFab& mf_dst, const int dcomp) const
```

File: Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.H [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+PartPerGridFunctor : public ComputeDiagFunctor
+{
+public:
+    /** Constructor.
+     * \param[in] mf_src source multifab. Must be nullptr as not source MF is needed
+     *            to compute the number of particles per cell, banane.
+     * \param[in] lev level of multifab. Used for averaging in rz.
+     * \param[in] ncomp Number of component of mf_src to cell-center in dst multifab.
+     */
+    PartPerGridFunctor(amrex::MultiFab* mf_src, int lev, int ncomp=1);
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.H]:
```suggestion
    PartPerGridFunctor(constamrex::MultiFab* mf_src, const int lev, const int ncomp=1);
```

File: Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.H [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+    /** \brief Compute the number of particles per cell directly into mf_dst.
+     *
+     * \param[out] mf_dst output MultiFab where the result is written
+     * \param[in] dcomp first component of mf_dst in which cell-centered
+     *            data is stored
+     */
+    virtual void operator()(amrex::MultiFab& mf_dst, int dcomp) const override final;
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.H]:
```suggestion
    virtual void operator()(amrex::MultiFab& mf_dst, const int dcomp) const override final;
```

File: Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.cpp [hunk 14]
```diff
@@ -0,0 +1,28 @@
+#include "PartPerGridFunctor.H"
+#include "Utils/Average.H"
+
+using namespace amrex;
+
+PartPerGridFunctor::PartPerGridFunctor(amrex::MultiFab* mf_src, int lev, int ncomp)
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.cpp]:
```suggestion
PartPerGridFunctor::PartPerGridFunctor(const amrex::MultiFab* mf_src, const int lev, const int ncomp)
```

File: Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.cpp [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    : ComputeDiagFunctor(ncomp), m_lev(lev)
+{
+    // mf_src will not be used, let's make sure it's null.
+    AMREX_ALWAYS_ASSERT(mf_src == nullptr);
+    // Write only in one output component.
+    AMREX_ALWAYS_ASSERT(ncomp == 1);
+}
+
+void
+PartPerGridFunctor::operator()(amrex::MultiFab& mf_dst, int dcomp) const
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/PartPerGridFunctor.cpp]:
```suggestion
PartPerGridFunctor::operator()(amrex::MultiFab& mf_dst, const int dcomp) const
```

File: Source/WarpX.H [hunk 16]
```diff
@@ -187,6 +187,13 @@ public:
     // do nodal
     static int do_nodal;
 
+    std::array<const amrex::MultiFab*, 3> get_array_Bfield_aux  (int lev) const { return {Bfield_aux[lev][0].get(),
+                                                                                          Bfield_aux[lev][1].get(),
+                                                                                          Bfield_aux[lev][2].get()}; }
+    std::array<const amrex::MultiFab*, 3> get_array_Efield_aux  (int lev) const { return {Efield_aux[lev][0].get(),
+                                                                                          Efield_aux[lev][1].get(),
+                                                                                          Efield_aux[lev][2].get()}; }
```
Comment (ax3l) [inline Source/WarpX.H]: These lines are a bit too long to read well on GitHub for reviews:
```suggestion
    std::array<const amrex::MultiFab*, 3>
    get_array_Bfield_aux(const int lev) const
    {
        return {
            Bfield_aux[lev][0].get(),
            Bfield_aux[lev][1].get(),
            Bfield_aux[lev][2].get()
        };
    }

    std::array<const amrex::MultiFab*, 3>
    get_array_Efield_aux(const int lev) const
    {
        return {
            Efield_aux[lev][0].get(),
            Efield_aux[lev][1].get(),
            Efield_aux[lev][2].get()
        };
    }
```

Maybe we should also add a doxygen string here?

File: Source/WarpX.H [hunk 17]
```diff
@@ -450,6 +457,8 @@ public:
                              const std::array<const amrex::MultiFab*, 3>& B,
                              const std::array<amrex::Real,3>& dx, int ngrow);
 
+    void ComputeDivE(amrex::MultiFab& divE, int lev);
```
Comment (ax3l) [inline Source/WarpX.H]:
```suggestion
    void ComputeDivE(amrex::MultiFab& divE, const int lev);
```

File: Source/WarpX.cpp [hunk 18]
```diff
@@ -1253,6 +1261,16 @@ WarpX::ComputeDivB (amrex::MultiFab& divB, int dcomp,
     }
 }
 
+void
+WarpX::ComputeDivE(amrex::MultiFab& divE, int lev)
```
Comment (ax3l) [inline Source/WarpX.cpp]:
```suggestion
WarpX::ComputeDivE(amrex::MultiFab& divE, const int lev)
```

File: Source/WarpX.H (hunk 16 above)
Comment (MaxThevenet) [reply Source/WarpX.H]: Ok I'll try some indentation and add a docstring.

**Comment by MaxThevenet** [comment]
Thanks! Just added a bunch of `const`.

**Pull request merged**